packages/ms-request-api/__tests__/shared/msRequestApiAgentLogic.ts

```diff
@@ -0,0 +1,88 @@
+import * as MsAuthenticator from '@sphereon/ms-authenticator'
+import { fetchIssuanceRequestMs } from '../../src/IssuerUtil'
+import { IMsRequestApi, IIssueRequest, IIssueRequestResponse } from '../../src/types/IMsRequestApi'
+import { TAgent } from '@veramo/core'
+import { v4 as uuidv4 } from 'uuid'
+
+type ConfiguredAgent = TAgent<IMsRequestApi>
+var requestIssuanceResponse: IIssueRequestResponse = {
+  requestId: '2e5c6fae-218c-4c7b-8440-df5454f908e9',
+  url: 'www.google.com',
+  expiry: new Date(1655935606),
+  id: 'fbef933e-f786-4b85-b1c8-6679346dc55d',
+  pin: '3683'
+
+}
+
+export default (testContext: { getAgent: () => ConfiguredAgent; setup: () => Promise<boolean>; tearDown: () => Promise<boolean> }) => {
+  describe('@sphereon/ms-request-api', () => {
+    let agent: ConfiguredAgent
+
+    beforeAll(async () => {
+
+      jest.mock('../../src/IssuerUtil', () => {
+        return {
+          fetchIssuanceRequestMs: jest.fn().mockResolvedValue(requestIssuanceResponse),
+          generatePin: jest.fn().mockResolvedValue(6363)
+        }
+      })
+
+      jest.mock('@sphereon/ms-authenticator', () => {
+        return {
+          ClientCredentialAuthenticator: jest.fn().mockResolvedValue('ey...'),
+          checkMsIdentityHostname: jest.fn().mockResolvedValue(MsAuthenticator.MS_IDENTITY_HOST_NAME_EU)
+        }
+      })
+      await testContext.setup()
+      agent = testContext.getAgent()
+
+    })
+
+    afterAll(async () => {
+      await new Promise<void>((resolve) => setTimeout(() => resolve(), 10000)) // avoid jest open handle error
+      await testContext.tearDown()
+    })
+
+    it('should request issuance from Issuer', async () => {
+      var requestConfigFile = '../../config/issuance_request_config.json';
+      var issuanceConfig = require(requestConfigFile);
+      var issuanceRequest: IIssueRequest = {
+        authenticationInfo: {
+          azClientId: 'AzClientID',
+          azClientSecret: 'AzClientSecret',
+          azTenantId: 'AzTenantId',
+          credentialManifestUrl: 'CredentialManifestUrl'
+        },
+        issuanceConfig: issuanceConfig
+      }
+
+      // modify the callback method to make it easier to debug
+      // with tools like ngrok since the URI changes all the time
+      // this way you don't need to modify the callback URL in the payload every time
+      // ngrok changes the URI
+      issuanceRequest.issuanceConfig.callback.url = `https://6270-2a02-a458-e71a-1-68b4-31d2-b44f-12b.eu.ngrok.io/api/issuer/issuance-request-callback`;
+
+      issuanceRequest.issuanceConfig.registration.clientName = "Sphereon Node.js SDK API Issuer";
+
+      // modify payload with new state, the state is used to be able to update the UI when callbacks are received from the VC Service
+      var id = uuidv4();
+      issuanceRequest.issuanceConfig.callback.state = id;
+
+      // here you could change the payload manifest
+      issuanceRequest.issuanceConfig.issuance.claims = {
```

Comment:
Can we move the claims out of the config? These are dyanmic per invocation. It doesn't make sense to me to have these as part of a config object

Reply:
Yes you are right, but Microsoft accepts claims(Credential Subject) as a part of issuance object. We can move the claims to out of config but during calling MS we have to put it to under request-json-> issuance-> claims pf MS Request API. It is possible to move it but then there will be very similar request objects one for client one for MS request which starts to be confusing. In my opinion, it is better to keep it like this to avoid complexity of code to have less reqest/response dtos for the same functionality. But if you think otherwise I can change.  

Sample request: 
{
  "authority": "did:ion:...0",
  "includeQRCode": false,
  "registration": {
    "clientName": "Verifiable Credential Expert Sample"
  },
  "callback": {
    "url": "https://7d93-2a02-a458-e71a-1-c4bf-628e-cc16-cff.eu.ngrok.io/api/issuer/issuanceCallback",
    "state": "...set at runtime...",
    "headers": {
      "api-key": "OPTIONAL API-KEY for ISSUANCE CALLBACK API"
    }
  },
  "issuance": {
    "type": "VerifiedCredentialExpert",
    "manifest": "https://beta.eu.did.msidentity.com/v1.0/e2a42b2f-7460-4499-afc2-425315ef058a/verifiableCredential/contracts/VerifiedCredentialExpert2",
    "pin": {
      "value": "123456",
      "length": 4
    },
    "claims": {
      "given_name": "FIRSTNAME",
      "family_name": "LASTNAME"
    }
  }
}

Reply:
The methods should be compatible with IPluginMethod which accepts only one object. So we can not send 2 objects as parameters. 
`
export interface IPluginMethod {
  (args: any, context: any): Promise<any>
}
`
But we can have 2 objects named IIssueRequest to IClientIssueRequest and convert IClientIssueRequest  to IIssueRequest at runtime. But this brings a bit more confusion/complexity in terms of development/developers.  Based on your concerns regarding to  end-user I will create a new object IClientIssueRequest like in the following and convert it to IIssueRequest at runtime.

`export interface IClientIssueRequest {
  authenticationInfo: IMsAuthenticationClientCredentialArgs
  issuanceConfig: IClientIssuanceConfig
  claims: CredentialSubject
}
`


`export interface IIssueRequest {
  authenticationInfo: IMsAuthenticationClientCredentialArgs
  issuanceConfig: IssuanceConfig
}
`

Reply:
This is better. And as long as we document it, it shouldn't be hard for external developers :)